File: rust/src/runtime/eval.rs
//! `judge()` — one-shot quality scoring against a goal + criteria.
//!
//! Rust port of node/src/runtime/eval.ts. Wraps the kernel's `gen_eval` free functions
//! (re-exported from `deepstrike_core::harness::eval`) into a small surface for callers
//! that just want "does this result meet the criteria?" without standing up
//! `HarnessLoop`.

use deepstrike_core::context::renderer::RenderedContext;
use deepstrike_core::harness::eval::{
    build_eval_messages as core_build_eval_messages, parse_verdict as core_parse_verdict,
    verdict_output_schema as core_verdict_output_schema, EvalResult,
};
use deepstrike_core::types::message::{Content, Message, Role};
use futures::StreamExt;

use crate::providers::{LLMProvider, StreamEvent};
use crate::Result;

/// Re-export the kernel's `Criterion` so callers don't need to import from `deepstrike_core`.
pub use deepstrike_core::harness::eval::Criterion;

/// The verdict shape parsed from the judge LLM. Same struct as the kernel's `EvalResult`.
pub type Verdict = EvalResult;

/// Build the kernel's eval prompt for (goal, criteria, result). Pure — does not call an LLM.
pub fn build_eval_messages(goal: &str, criteria: &[Criterion], result: &str) -> Vec<Message> {
    core_build_eval_messages(goal, criteria, result, 1, false)
}

/// Parse a Verdict from raw judge-LLM text.
pub fn parse_verdict(text: &str) -> Verdict {
    core_parse_verdict(text)
}

/// The JSON Schema the kernel expects judge output to conform to.
pub fn verdict_output_schema() -> serde_json::Value {
    core_verdict_output_schema(false)
}

/// One-shot judge: render the eval prompt, stream the provider, parse the verdict.
/// Errors when the provider returns no text.
pub async fn judge(
    provider: &dyn LLMProvider,
    goal: &str,
    criteria: &[Criterion],
    result: &str,
) -> Result<Verdict> {
    let msgs = build_eval_messages(goal, criteria, result);
    let system_text: String = msgs
        .iter()
        .filter(|m| m.role == Role::System)
        .filter_map(|m| match &m.content {
            Content::Text(s) => Some(s.clone()),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join("\n\n");
    let turns: Vec<Message> = msgs.into_iter().filter(|m| m.role != Role::System).collect();
    let ctx = RenderedContext {
        system_text,
        system_stable: String::new(),
        system_knowledge: String::new(),
        state_turn: None,
        turns,
        frozen_prefix_len: None,
    };

    let mut text = String::new();
    let mut stream = provider.stream(&ctx, &[], None, None).await?;
    while let Some(evt) = stream.next().await {
        if let Ok(StreamEvent::TextDelta { delta }) = evt {
            text.push_str(&delta);
        }
    }
    if text.is_empty() {
        return Err(crate::Error::Other("judge: provider produced no text".to_string()));
    }
    Ok(parse_verdict(&text))
}

File: node/src/runtime/eval.ts
/**
 * `judge()` — one-shot quality scoring against a goal + criteria using the kernel's `gen_eval`.
 *
 * Wraps the three kernel free functions `buildEvalMessages` / `parseVerdict` / `verdictOutputSchema`
 * (folded out of the old EvalPipeline class in 0.5.0) into a small typed surface that's safe to
 * call from a benchmark harness, a CI gate, or any caller that just wants "does this result meet
 * the criteria?" without setting up `AttemptLoop`.
 *
 * The judge is a single LLM call: build the eval prompt → stream → parse verdict. No retry loop,
 * no skill extraction, no loop state. Use `AttemptLoop` if you want the retry/refine flow.
 */

import type { LLMProvider, Message, RenderedContext, TextDelta } from "../types.js"
import { getKernel } from "../kernel.js"

export interface Criterion {
  /** The criterion text the judge evaluates against. */
  text: string
  /** When true (default), failing this criterion fails the overall verdict. */
  required?: boolean
  /** Optional weight for weighted scoring (kernel-defined semantics). */
  weight?: number
  /** Stable host contract identifier, passed through to deterministic judges. */
  id?: string
  /** Signals that the host can evaluate this criterion without an LLM. */
  machineCheckable?: boolean
}

export interface VerdictDetail {
  criterion: string
  passed: boolean
  score: number
  feedback: string
}

export interface Verdict {
  passed: boolean
  /** 0..1 — kernel-defined aggregate score. */
  overallScore: number
  feedback: string
  details: VerdictDetail[]
}

export interface JudgeArgs {
  /** Provider used for the eval LLM call. Often a cheaper model than the main run. */
  provider: LLMProvider
  /** The task goal the result is being evaluated against. */
  goal: string
  /** The criteria the judge scores against. */
  criteria: Criterion[]
  /** The agent's result text (final reply, or a structured summary when the run was incomplete). */
  result: string
  /** Optional abort signal forwarded to provider.stream. */
  signal?: AbortSignal
}

/**
 * Build the kernel's eval prompt for (goal, criteria, result).
 * Exposed in case a caller wants to render the prompt without calling the LLM (e.g., dry-run cost
 * estimation, fixture generation). For the common case, use `judge()`.
 */
export function buildEvalMessages(goal: string, criteria: Criterion[], result: string): Message[] {
  return getKernel().buildEvalMessages(
    goal,
    criteria.map(c => ({ text: c.text, required: c.required ?? true, weight: c.weight })),
    result,
    1,     // attempt
    false, // extractSkillOnPass — not relevant for benchmark scoring
  )
}

/** Parse a Verdict from raw judge-LLM text. Throws on schema mismatch. */
export function parseVerdict(text: string): Verdict {
  const v = getKernel().parseVerdict(text)
  return {
    passed: v.passed,
    overallScore: v.overallScore,
    feedback: v.feedback,
    details: v.details ?? [],
  }
}

/** The JSON Schema the kernel expects judge output to conform to. */
export function verdictOutputSchema(): Record<string, unknown> {
  return JSON.parse(getKernel().verdictOutputSchema(false)) as Record<string, unknown>
}

/**
 * Run one judge pass: render the eval prompt, stream the provider, parse the verdict.
 * Throws when the provider returns no text or returns content that fails verdict parsing.
 */
export async function judge(args: JudgeArgs): Promise<Verdict> {
  const msgs = buildEvalMessages(args.goal, args.criteria, args.result)
  const ctx: RenderedContext = {
    systemText: msgs.filter(m => m.role === "system").map(m => m.content).join("\n\n"),
    turns: msgs.filter(m => m.role !== "system"),
  }
  let text = ""
  for await (const evt of args.provider.stream(ctx, [], undefined, undefined, args.signal)) {
    if (evt.type === "text_delta") text += (evt as TextDelta).delta
  }
  if (!text) {
    throw new Error("judge: provider produced no text")
  }
  return parseVerdict(text)
}
